xonsh/lexer.py

```diff
@@ -1,13 +1,307 @@
+            found_match = True
+            break
+        try:
+            n = next(stream)
+        except:
+            n = None
+    if found_match:
+        yield _new_token('REGEXPATH', sofar, token.start)
+        state['last'] = n
+    else:
+        e = "Could not find matching backtick for regex on line {0}"
+        m = e.format(token.start[0])
+        yield _new_token("ERRORTOKEN", m, token.start)
+
+
+def handle_newline(state, token, stream):
+    try:
+        n = next(stream)
+    except:
+        n = None
+
+    yield _new_token('NEWLINE', '\n', token.start)
+    state['last'] = token
+
+    if n is not None:
+        if n.type != tokenize.ENDMARKER:
+            for i in handle_token(state, n, stream):
+                yield i
+
+
+def handle_lparen(state, token, stream):
+    state['pymode'].append(True)
+    state['last'] = token
+    yield _new_token('LPAREN', '(', token.start)
+
+
+def handle_lbrace(state, token, stream):
+    state['pymode'].append(True)
+    state['last'] = token
+    yield _new_token('LBRACE', '{', token.start)
+
+
+def handle_lbracket(state, token, stream):
+    state['pymode'].append(True)
+    state['last'] = token
+    yield _new_token('LBRACKET', '[', token.start)
+
+
+def handle_rparen(state, token, stream):
+    state['pymode'].pop()
+    state['last'] = token
+    yield _new_token('RPAREN', ')', token.start)
+
+
+def handle_rbrace(state, token, stream):
+    state['pymode'].pop()
+    state['last'] = token
+    yield _new_token('RBRACE', '}', token.start)
+
+
+def handle_rbracket(state, token, stream):
+    state['pymode'].pop()
+    state['last'] = token
+    yield _new_token('RBRACKET', ']', token.start)
+
+
+def handle_error_space(state, token, stream):
+    if not state['pymode'][-1]:
+        state['last'] = token
+        yield _new_token('WS', ' ', token.start)
+    else:
+        yield from []
+
+special_handlers = {
+    tokenize.ENCODING: lambda s, t, st: [],
+    tokenize.COMMENT: lambda s, t, st: [],
+    tokenize.ENDMARKER: lambda s, t, st: [],
+    tokenize.NEWLINE: handle_newline,
+    (tokenize.OP, '('): handle_lparen,
+    (tokenize.OP, ')'): handle_rparen,
+    (tokenize.OP, '['): handle_lbracket,
+    (tokenize.OP, ']'): handle_rbracket,
+    (tokenize.OP, '{'): handle_lbrace,
+    (tokenize.OP, '}'): handle_rbrace,
+    (tokenize.ERRORTOKEN, '$'): handle_dollar,
+    (tokenize.ERRORTOKEN, '`'): handle_backtick,
+    (tokenize.ERRORTOKEN, '?'): handle_question,
+    (tokenize.OP, '@'): handle_at,
+    (tokenize.ERRORTOKEN, ' '): handle_error_space,
+    tokenize.INDENT: handle_indent,
+    tokenize.DEDENT: handle_indent,
+}
+
+
+def handle_token(state, token, stream):
+    typ = token.type
+    st = token.string
+    if not state['pymode'][-1]:
+        if state['last'] is not None and state['last'].end != token.start:
+            cur = token.start
+            old = state['last'].end
+            yield _new_token('WS', ' '*(cur[1]-old[1]), old)
```

Comment:
I think that this could cause some weird issues when debugging, thought the column count should still be correct.  I don't really want for this issue to block the whole PR, though.  Maybe you could make an issue for it so that we remember to investigate later?